pkg/controllermanager/controller/project/project_control_reconcile.go

```diff
@@ -216,3 +247,48 @@ func (c *defaultControl) reconcileNamespaceForProject(project *gardencorev1beta1
 
 	return namespace, nil
 }
+
+// CustomProjectRole is a constant for a label value.
+const CustomProjectRole = "custom-project-role"
+
+func deleteStaleCustomRoles(ctx context.Context, c client.Client, nonStaleCustomRoles sets.String, namespace string) error {
+	for _, list := range []runtime.Object{
+		&rbacv1.RoleBindingList{},
+		&rbacv1.ClusterRoleList{},
+	} {
+		if err := c.List(
+			ctx,
+			list,
+			client.InNamespace(namespace),
+			client.MatchingLabels{
+				v1beta1constants.GardenRole: CustomProjectRole,
+			},
+		); err != nil {
+			return err
+		}
+
+		if err := meta.EachListItem(list, func(obj runtime.Object) error {
+			accessor, err := meta.Accessor(obj)
+			if err != nil {
+				return err
+			}
+
+			if nonStaleCustomRoles.Has(getCustomRoleNameFromRBAC(accessor.GetName())) {
+				return nil
+			}
+
+			return client.IgnoreNotFound(c.Delete(ctx, obj))
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getCustomRoleNameFromRBAC(name string) string {
+	if i := strings.LastIndex(name, ":"); i != -1 {
```

Comment:
This means, custom roles may not contain `:`, right?
I think we should rather do something like `strings.TrimPrefix(name, "gardener.cloud:custom:project:<project-name>:")`